### PR TITLE
Migrate Donation model to the new Address data structure and improve addresses management

### PR DESCRIPTION
**Pull request by tgalopin**

### CODE REVIEW

File: src/AppBundle/Validator/RecaptchaValidator.php [hunk 1]
```diff
@@ -16,21 +16,21 @@ public function __construct(RecaptchaApiClient $httpClient)
         $this->httpClient = $httpClient;
     }
 
-    public function validate($value, Constraint $constraint)
+    public function validate($address, Constraint $constraint)
```
Comment (HeahDude) [inline src/AppBundle/Validator/RecaptchaValidator.php]: Why this change?

File: src/AppBundle/Validator/UniqueCommitteeValidator.php [hunk 2]
```diff
@@ -18,23 +18,23 @@ public function __construct(CommitteeRepository $repository)
         $this->repository = $repository;
     }
 
-    public function validate($value, Constraint $constraint)
+    public function validate($address, Constraint $constraint)
```
Comment (HeahDude) [inline src/AppBundle/Validator/UniqueCommitteeValidator.php]: Why?

File: src/AppBundle/Validator/UniqueMembershipValidator.php [hunk 3]
```diff
@@ -18,23 +18,23 @@ public function __construct(AdherentRepository $repository)
         $this->repository = $repository;
     }
 
-    public function validate($value, Constraint $constraint)
+    public function validate($address, Constraint $constraint)
```
Comment (HeahDude) [inline src/AppBundle/Validator/UniqueMembershipValidator.php]: This change looks wrong to me.

File: src/AppBundle/Validator/UnitedNationsCountryValidator.php [hunk 4]
```diff
@@ -15,18 +15,18 @@ class UnitedNationsCountryValidator extends ConstraintValidator
     /**
      * {@inheritdoc}
      */
-    public function validate($value, Constraint $constraint)
+    public function validate($address, Constraint $constraint)
```
Comment (HeahDude) [inline src/AppBundle/Validator/UnitedNationsCountryValidator.php]: Same here.

File: src/AppBundle/Validator/ValidAddressValidator.php [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+use AppBundle\Intl\FranceCitiesBundle;
+use Symfony\Component\PropertyAccess\PropertyAccessor;
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\ConstraintValidator;
+use Symfony\Component\Validator\Exception\UnexpectedTypeException;
+
+/**
+ * @Annotation
+ */
+class ValidAddressValidator extends ConstraintValidator
```
Comment (HeahDude) [inline src/AppBundle/Validator/ValidAddressValidator.php]: `AddressValidator`.

File: src/AppBundle/Validator/ValidAddressValidator.php [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+{
+    private $propertyAccessor;
```
Comment (HeahDude) [inline src/AppBundle/Validator/ValidAddressValidator.php]: What for?

File: src/AppBundle/Validator/WasNotInvitedRecentlyValidator.php [hunk 7]
```diff
@@ -24,13 +24,13 @@ public function __construct(EntityManager $entityManager, PropertyAccessor $prop
     /**
      * {@inheritdoc}
      */
-    public function validate($value, Constraint $constraint)
+    public function validate($address, Constraint $constraint)
```
Comment (HeahDude) [inline src/AppBundle/Validator/WasNotInvitedRecentlyValidator.php]: Please revert all these changes.

File: tests/AppBundle/Controller/AdherentControllerTest.php [hunk 8]
```diff
@@ -110,14 +110,13 @@ public function testEditAdherentProfile()
         $errors = $crawler->filter('.form__errors > li');
 
         $this->assertResponseStatusCode(Response::HTTP_OK, $this->client->getResponse());
-        $this->assertSame(7, $errors->count());
+        $this->assertSame(6, $errors->count());
         $this->assertSame('Cette valeur ne doit pas être vide.', $errors->eq(0)->text());
         $this->assertSame('Cette valeur ne doit pas être vide.', $errors->eq(1)->text());
-        $this->assertSame('Cette ville et ce code postal ne sont pas liés.', $errors->eq(2)->text());
+        $this->assertSame('Cette valeur n\'est pas un code postal français valide.', $errors->eq(2)->text());
```
Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Should be reverted too IMO.

File: tests/AppBundle/Controller/AdherentControllerTest.php [hunk 9]
```diff
@@ -371,10 +370,9 @@ public function testRegularAdherentCanCreateOneNewCommittee(string $emaiLAddress
         ]));
 
         $this->assertResponseStatusCode(Response::HTTP_OK, $this->client->getResponse());
-        $this->assertSame(11, $crawler->filter('#create-committee-form .form__errors > li')->count());
-        $this->assertSame('Cette ville et ce code postal ne sont pas liés.', $crawler->filter('#committee-address > .form__errors > .form__error')->eq(0)->text());
+        $this->assertSame(10, $crawler->filter('#create-committee-form .form__errors > li')->count());
+        $this->assertSame('Cette valeur n\'est pas un code postal français valide.', $crawler->filter('#committee-address > .form__errors > .form__error')->eq(0)->text());
```
Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Same here.

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 8 above)
Comment (tgalopin) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Why ?

Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Because the value is hidden to the end user, to him the zip code will be right, but the link to a city is broken.

File: app/Resources/views/donation/index.html.twig [hunk 10]
```diff
@@ -62,29 +62,32 @@
...
+                        {{ form_errors(form.address.cityName) }}
+                        {{ form_widget(form.address.cityName, { 'attr': { 'class': 'form--full', 'placeholder': 'Ville' } }) }}
+                        {{ form_widget(form.address.city) }}
                     </div>
                 </div>
 
                 <div class="form__row">
-                    {{ form_errors(form.address) }}
-                    {{ form_widget(form.address, { 'attr': { 'class': 'form--full', 'placeholder': 'Adresse postale' } }) }}
+                    {{ form_errors(form.address.address) }}
```
Comment (tgalopin) [inline app/Resources/views/donation/index.html.twig]: I think this should be another PR as the DTO is used in other places.

File: src/AppBundle/Controller/DonationController.php [hunk 11]
```diff
@@ -22,26 +22,30 @@ class DonationController extends Controller
...
             'countries' => UnitedNationsBundle::getCountries($request->getLocale()),
         ]);
     }
 
     /**
-     * @Route("/{id}/paiement", name="donation_pay", requirements={"id"="^[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$"})
+     * @Route(
+     *     "/{uuid}/paiement",
+     *     name="donation_pay",
+     *     requirements={"uuid"="^[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{4}-[0-9A-Fa-f]{12}$"}
```
Comment (tgalopin) [inline src/AppBundle/Controller/DonationController.php]: What do you mean?

Comment (HeahDude) [inline src/AppBundle/Controller/DonationController.php]: https://github.com/EnMarche/en-marche.fr/blob/master/src/AppBundle/Controller/CommitteeController.php#L19

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 8 above)
Comment (tgalopin) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: I changed the behavior of the rule as it was wrong: we wern't checking properly the country. This is the new expected behavior.

File: src/AppBundle/Validator/ValidAddressValidator.php (hunk 5 above)
Comment (tgalopin) [inline src/AppBundle/Validator/ValidAddressValidator.php]: I prefer to have `ValidAddress` and `GeocodableAddress` constraints and to group constraints with validators. What do you propose?

Comment (HeahDude) [inline src/AppBundle/Validator/ValidAddressValidator.php]: The alias in the use statement does not have to be named as the constraint class. Naming the constraint this way implies you only validate valid address, using `AssertValidAddress` as alias for the annotation makes more sense.

File: tests/AppBundle/Controller/AdherentControllerTest.php (hunk 8 above)
Comment (HeahDude) [inline tests/AppBundle/Controller/AdherentControllerTest.php]: Ok then.

File: src/AppBundle/Donation/DonationRequest.php [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+    {
+        return $this->lastName;
+    }
+
+    public function setLastName(?string $lastName)
+    {
+        $this->lastName = $lastName;
+    }
+
+    public function setAddress(Address $address = null)
```
Comment (HeahDude) [inline src/AppBundle/Donation/DonationRequest.php]: `?Address`

File: src/AppBundle/Donation/DonationRequestFactory.php [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            $donation->setPhone($phone);
+        }
+
+        return $donation;
+    }
+
+    public function createFromAdherent(Adherent $adherent, int $defaultAmount = 50): DonationRequest
+    {
+        $donation = DonationRequest::createFromAdherent($adherent);
+        $donation->setAmount($defaultAmount);
```
Comment (HeahDude) [inline src/AppBundle/Donation/DonationRequestFactory.php]: You should really pass amount as constructor and static method argument to avoid all these setter calls.
